### PR TITLE
Added filters get resources

### PR DESCRIPTION
**Pull request by alsastre**
Could not be fully tested because docker image conjur-appliance-cuke-master was not available.
It has been tested manually against a local conjur instance.

### CODE REVIEW

**Comment by doodlesbykumbi** [comment]
@alsastre Thanks for kicking off this PR. 

This should have kicked off a gitlab pipeline, let me look into what's happening there.

In the meantime, could you please add test cases for Limit, Offset and Count ?

File: conjurapi/resource.go [hunk 1]
```diff
@@ -75,6 +79,14 @@ func (c *Client) Resources(filter *ResourceFilter) (resources []map[string]inter
 	}
 
 	resources = make([]map[string]interface{}, 1)
-	err = json.Unmarshal(data, &resources)
+
+	// When count is enabled, the result is not an array
+	if filter.Count {
```
Comment (doodlesbykumbi) [inline conjurapi/resource.go]: Let's add a `nil` check here otherwise this panics when `filter == nil`.

File: conjurapi/resource_test.go [hunk 2]
```diff
@@ -110,6 +110,7 @@ func TestClient_Resources(t *testing.T) {
 
 		Convey("Lists all resources", listResources(conjur, nil))
 		Convey("Lists resources by kind", listResources(conjur, &ResourceFilter{Kind: "variable"}))
+		Convey("Lists resources that start with db", listResources(conjur, &ResourceFilter{Search: "db"}))
```
Comment (doodlesbykumbi) [inline conjurapi/resource_test.go]: Let's add test cases for Limit, Offset and Count

File: conjurapi/resource_test.go [hunk 3]
```diff
@@ -96,20 +98,24 @@ func TestClient_CheckPermission(t *testing.T) {
 }
 
 func TestClient_Resources(t *testing.T) {
-	listResources := func(conjur *Client, filter *ResourceFilter) func() {
+	listResources := func(conjur *Client, filter *ResourceFilter, int expected) func() {
```
Comment (doodlesbykumbi) [inline conjurapi/resource_test.go]: `expected int` ?

File: conjurapi/resource.go [hunk 4]
```diff
@@ -75,6 +79,17 @@ func (c *Client) Resources(filter *ResourceFilter) (resources []map[string]inter
 	}
 
 	resources = make([]map[string]interface{}, 1)
-	err = json.Unmarshal(data, &resources)
+
+	// When count is enabled, the result is not an array
+	// so we have to add the result to an array to match the resulting type
+	if filter != nil {
```
Comment (doodlesbykumbi) [inline conjurapi/resource.go]: Currently with `filter == nil` this func returns empty resources, this isn't the expected behaviour.
The outer `if` isn't necessary.
```
if filter != nil && filter.Count {
	value := make(map[string]interface{})
	err = json.Unmarshal(data, &value)
	resources[0] = value
} else {
	err = json.Unmarshal(data, &resources)
}
```

Comment (alsastre) [reply conjurapi/resource.go]: You are right, made the changes

File: conjurapi/resource_test.go (hunk 3 above)
Comment (alsastre) [reply conjurapi/resource_test.go]: Can't test since the image `conjur-appliance-cuke-master` is not available hence `docker-compose` fails

Comment (doodlesbykumbi) [reply conjurapi/resource_test.go]: @alsastre sorry i should have been clearer - `expected int` as opposed to `int expected`, as in `func (arg_name type, ...)`

Comment (doodlesbykumbi) [reply conjurapi/resource_test.go]: `./test.sh oss` and `./dev.sh oss` should not require the use of cuke-master unless something is off

Comment (alsastre) [reply conjurapi/resource_test.go]:
```
(feature/new_filters_resources) ~/.../alsastre/conjur-api-go $ ./test.sh 
+ trap finish EXIT
+ echo 'Running tests'
Running tests
+ rm -rf output
+ mkdir -p output
+ source ./_setup.sh
+++ sed 's/^[_-]*\(.*\)/\1/'
+++ basename /home/alejandro/Dropbox/Projects/Go/src/github.com/alsastre/conjur-api-go
++ export COMPOSE_PROJECT_NAME=conjur-api-go
++ COMPOSE_PROJECT_NAME=conjur-api-go
++ export TEST_VERSION=all
++ TEST_VERSION=all
++ oss_only
++ '[' all == oss ']'
++ docker-compose pull conjur cuke-master
Pulling conjur      ... done
Pulling cuke-master ... error

ERROR: for cuke-master  Get https://registry2.itci.conjur.net/v2/conjur-appliance-cuke-master/manifests/4.9-stable: unauthorized: authentication required
ERROR: Get https://registry2.itci.conjur.net/v2/conjur-appliance-cuke-master/manifests/4.9-stable: unauthorized: authentication required
+ finish
+ docker-compose down -v
Removing network conjur-api-go_default
WARNING: Network conjur-api-go_default not found.
```

Comment (alsastre) [reply conjurapi/resource_test.go]: > @alsastre sorry i should have been clearer - `expected int` as opposed to `int expected`, as in `func (arg_name type, ...)`

:+1: Changed. Problem of changing between python and go

Comment (alsastre) [reply conjurapi/resource_test.go]: > `./test.sh oss` and `./dev.sh oss` should not require the use of cuke-master unless something is off

Looking at dev and test,, both load `_setup.sh` the variable `TEST_VERSION=` was not corrreclty set. After forcing it to be `oss` I am able to test. 
AFTER TESTING:
Have fixed some errors and now it passes all the test

Comment (alsastre) [reply conjurapi/resource_test.go]: Changing the following in test.sh allows the test to work: 
```
-source ./_setup.sh
+source ./_setup.sh oss
 ```
I suppose `oss` could be substituted by `$1` , and with this change ` test.sh` will work with community if it is called with the argument `oss`.

Comment (doodlesbykumbi) [reply conjurapi/resource_test.go]: @alsastre `./test.sh oss`, not `./test.sh`. you'll have to add the `oss` argument to avoid cuke-master

Comment (doodlesbykumbi) [reply conjurapi/resource_test.go]: oh sorry just saw your comments above :)

Comment (doodlesbykumbi) [reply conjurapi/resource_test.go]: `_setup.sh` should pickup the argument though

Comment (alsastre) [reply conjurapi/resource_test.go]: I checked, it does pick it up. Its only that I didn't saw where. So no fix needed only a comment on documentation to explain it

**Comment by doodlesbykumbi** [comment]
Pinging @alsastre, are you able to address the above comment ?

**Comment by alsastre** [comment]
Yes.Will upload the new version hopefully this or next weekend

**Comment by alsastre** [comment]
I have been playing wiith it and I don't like much the idea of having a separate function. We are duplicating functionality in order to retrieve the count parsed directly as int. I think it wouldn't be much of a pain  to the user of the api to change the reception of the values depending on the filter sent to the api. I think that is better than replicating and limiting functionality in the api side. Because both methods are the same only that one only accepts Filters with Count to true and the other with Count to false.
I would even go as far to remove the count option because is much comfortable  for a user of the api to retrieve the variables and check the len of the resulting array than having to call a specific new method that only serves for counting but that at the same time also receives a Filter structure. 
My proposal: Remove Count from the filter, and let the user of the api use the  `len(Resources(filter))` when he only wants the number of items.

**Comment by alsastre** [comment]
@doodlesbykumbi Any decision related to this?

**Comment by sgnn7** [comment]
@alsastre Thank you for the PR and sorry for the delays.

I looked at the code in the meantime and I also agree with @doodlesbykumbi that the `Count bool` as a parameter does seem like a really unintuitive way to request the count of items, especially if that field is not really compatible with the rest of the param combinations so maybe a new method or removal of the count would be the best path forward.

As a sidenote on the actual code, I think it would be valuable to use [Values struct](https://golang.org/pkg/net/url/#Values.Add) as the primary way of chaining query params.

Hopefully that helps!

**Comment by alsastre** [comment]
I have removed the Count from the filter since I think that anyone that wants to use Count can use length method afterwards.
@sgnn7 Thank you for the Values struct, it does look clearer now.

**Comment by sgnn7** [comment]
@alsastre Awesome! 

Since the PR had a bit of a messy git history in regards to fixups and long descriptions and we don't need to keep going back/forth more on this, I'm going to squash-merge the PR. Thanks again!

**Pull request merged**